code/components/devtools-five/src/ArchetypeDebug.cpp

```diff
@@ -0,0 +1,292 @@
+{
+public:
+	atMultiHashMap<T> data;
+};
+
+struct FactoryInfo
+{
+	FactoryInfo(const char* factoryName) : name(factoryName), count(0), usage(0.0f) {}
+
+	std::string name;
+
+	size_t count;
+
+	float usage;
+};
+
+struct FactoryBucket
+{
+	FactoryBucket() : count(0) {}
+
+	std::string name;
+
+	size_t count;
+};
+
+static void* rage__fwArchetypeManager__ms_ArchetypePool;
+
+// Taken from CloneExperiments.cpp
+static hook::cdecl_stub<const char* (int, uint32_t)> rage__atHashStringNamespaceSupport__GetString([]
+{
+	return hook::get_pattern("85 D2 75 04 33 C0 EB 61", -0x18);
+});
+
+static HookFunction hookFunction([]()
+{
+	auto location = hook::get_call(hook::get_pattern<char>("E8 ? ? ? ? EB 0B 45 33 C0 48 8B CB"));
+	rage__fwArchetypeManager__ms_ArchetypePool = hook::get_address<void*>(location + 0x2B);
+});
+
+static std::string ResolveHashString(uint32_t hash)
+{
+	const char* resolvedName = rage__atHashStringNamespaceSupport__GetString(1, hash);
+	if (resolvedName)
+	{
+		return resolvedName;
+	}
+
+	std::string fileName = streaming::GetStreamingBaseNameForHash(hash);
+	if (!fileName.empty())
+	{
+		return fileName;
+	}
+
+	return fmt::sprintf("Unknown 0x%08X", hash);
+}
+
+static std::string GetFriendlyNameFromBucketHash(uint32_t hash)
+{
+	if (hash == 0xFFFF)
+	{
+		return "Base Game Content";
+	}
+	else if (hash == 0xF000)
+	{
+		return "Other Game Content";
+	}
+	else if (hash < 0xFFFF)
+	{
+		static auto mapTypesStore = streaming::Manager::GetInstance()->moduleMgr.GetStreamingModule("ytyp");
+		void* mapTypesContent = mapTypesStore->GetPtr(hash);
+		if (mapTypesContent)
+		{
+			hash = *(uint32_t*)((char*)mapTypesContent + 0x8);
+		}
+	}
+	return ResolveHashString(hash);
+}
+
+template<typename T>
+static int SortCompare(const T& lhs, const T& rhs)
+{
+	if (lhs < rhs)
+	{
+		return -1;
+	}
+	else if (lhs > rhs)
+	{
+		return 1;
+	}
+	else
+	{
+		return 0;
+	}
+}
+
+static InitFunction initFunction([]()
+{
+	static bool archetypeListEnabled;
+	static ConVar<bool> archetypeListVar("archetypelist", ConVar_Archive | ConVar_UserPref, false, &archetypeListEnabled);
+
+	ConHost::OnShouldDrawGui.Connect([](bool* should)
+	{
+		*should = *should || archetypeListEnabled;
+	});
+
+	ConHost::OnDrawGui.Connect([]()
+	{
+		if (!archetypeListEnabled)
+		{
+			return;
+		}
+
+		if (g_archetypeFactories->GetSize() == 0 || g_archetypeFactories->GetCount() < 7)
+		{
+			return;
+		}
+
+		ImGui::SetNextWindowSize(ImVec2(500.0f, 300.0f), ImGuiCond_FirstUseEver);
+
+		std::vector<FactoryInfo> factoryInfos = {
+			"CBaseModelInfo",
+			"CMloModelInfo",
+			"CTimeModelInfo",
+			"CWeaponModelInfo",
+			"CVehicleModelInfo",
+			"CPedModelInfo",
+			"CCompEntityModelInfo",
+		};
+
+		std::unordered_map<uint32_t, FactoryBucket> factoryBucketsLookup;
+
+		size_t archetypePoolSize = *(uint32_t*)(((char*)rage__fwArchetypeManager__ms_ArchetypePool) + 0x8);
+		size_t archetypePoolFree = *(uint32_t*)(((char*)rage__fwArchetypeManager__ms_ArchetypePool) + 0x10);
```

Comment:
So our `atPoolBase` code completely lack support for raw values in "flags" (aka "free count" or so) field, we definitely need to improve this in the future.

Reply:
@Disquse can you make a task on this?